python/pytests/substring_test.py

```diff
@@ -0,0 +1,35 @@
+import kaskada as kd
+import pytest
+
+
+@pytest.fixture(scope="module")
+async def source() -> kd.sources.CsvString:
+    content = "\n".join(
+        [
+            "time,key,s,n,t",
+            '1996-12-19T16:39:57,A,"hEllo",0,"hEllo"',
+            '1996-12-19T16:39:58,B,"World",5,"world"',
+            '1996-12-19T16:39:59,B,"hello world",-2,"hello world"',
+            '1996-12-19T16:40:00,B,,-2,"greetings"',
+            '1996-12-19T16:40:01,B,,2,"salutations"',
+            '1996-12-19T16:40:02,B,"goodbye",,',
+        ]
+    )
+    return await kd.sources.CsvString.create(
+        content, time_column="time", key_column="key"
+    )
+
+
+async def test_substring(source, golden) -> None:
```

Comment:
We may also want to test with negative numbers (I think substring uses that to reflect "distance from the end"). So `start=-1` would capture the last character, etc.

Reply:
this already exists via line 12, 31, 32